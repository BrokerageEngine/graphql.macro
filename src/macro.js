// @flow
import path from 'path';
import { createMacro } from 'babel-plugin-macros';
import gqlTag from 'graphql-tag';
import serialize from 'babel-literal-to-ast';
import expandImports from './utils/expandImports';
// import printAST from 'ast-pretty-print';
// console.log(printAST(referencePath.parentPath))

function graphqlTagMacro({
  references,
  state: { file: { opts: { filename } } },
}: {
  references: { gql: Array<any>, loader: Array<any> },
  state: { file: { opts: { filename: string } } },
}): void {
  const { gql = [], loader = [] } = references;

  // Case 1: import { gql } from 'graphql-tag.macro'
  gql.forEach(referencePath => {
    const { raw } = referencePath.parentPath.node.quasi.quasis[0].value;
    referencePath.parentPath.replaceWith(serialize(gqlTag(raw)));
  });

  // Case 2: import { loader } from 'graphql-tag.macro'
  loader.forEach(referencePath => {
    referencePath.parentPath.node.arguments.forEach(({ value }) => {
      const queryPath = path.join(filename, '..', value);
      const expanded = expandImports(queryPath); // Note: #import feature
      referencePath.parentPath.replaceWith(serialize(gqlTag(expanded)));
    });
  });
}

export default createMacro(graphqlTagMacro);
