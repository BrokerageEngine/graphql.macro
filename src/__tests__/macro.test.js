import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';

expect.addSnapshotSerializer({
  print(val) {
    return val.replace(/..\/macro/, 'graphql-tag.macro');
  },
  test(val) {
    return typeof val === 'string';
  },
});

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    filename: __filename,
  },
  tests: {
    '[gql]': {
      error: false,
      code: `
        import { gql } from '../macro';
        const query = gql\`
          {
            user(id: 5) {
              firstName
              lastName
            }
          }
        \`;
      `,
    },
    '[loader]': {
      error: false,
      code: `
        import { loader } from '../macro';
        const query = loader('./fixtures/query1.graphql');
      `,
    },
    // '[loader] multiple operations': {
    //   error: false,
    //   code: `
    //     import { loader } from '../macro';
    //     const { User, Post } = loader('./fixtures/query2.graphql');
    //   `,
    // },
  },
});
