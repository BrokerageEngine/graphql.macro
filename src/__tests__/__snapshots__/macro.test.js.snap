// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros [gql]: [gql] 1`] = `

import { gql } from 'graphql-tag.macro';
const query = gql\`
  {
    user(id: 5) {
      firstName
      lastName
    }
  }
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "query",
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "user"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "value": {
            "kind": "IntValue",
            "value": "5"
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "firstName"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "lastName"
            },
            "arguments": [],
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 64,
    "source": {
      "body": "\\n  {\\n    user(id: 5) {\\n      firstName\\n      lastName\\n    }\\n  }\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;

exports[`macros [loader]: [loader] 1`] = `

import { loader } from 'graphql-tag.macro';
const query = loader('./fixtures/query1.graphql');

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry1"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "firstName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }, {
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry3"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "firstName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }, {
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry2"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "lastName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }, {
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "User"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "user"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "value": {
            "kind": "IntValue",
            "value": "5"
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "FragmentSpread",
            "name": {
              "kind": "Name",
              "value": "UserEntry1"
            },
            "directives": []
          }, {
            "kind": "FragmentSpread",
            "name": {
              "kind": "Name",
              "value": "UserEntry2"
            },
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 264,
    "source": {
      "body": "fragment UserEntry1 on User {\\n  firstName\\n}\\n\\nfragment UserEntry3 on User {\\n  firstName\\n}\\nfragment UserEntry2 on User {\\n  lastName\\n}\\n#import \\"./fragment1.graphql\\"\\n#import './fragment2.graphql'\\n\\nquery User {\\n  user(id: 5) {\\n    ...UserEntry1\\n    ...UserEntry2\\n  }\\n}\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;
